include/ghex/unstructured/pattern.hpp

```diff
@@ -0,0 +1,346 @@
+                struct extended_domain_id_type {
+
+                    // members
+                    domain_id_type id;
+                    int mpi_rank;
+                    address_type address;
+                    int tag;
+
+                    // member functions
+                    // /** @brief unique ordering given by id and tag*/
+                    // bool operator < (const extended_domain_id_type& other) const noexcept {
+                    //     return (id < other.id ? true : (id == other.id ? (tag < other.tag) : false));
+                    // }
+                    /** @brief unique ordering given by address and tag*/
+                    bool operator < (const extended_domain_id_type& other) const noexcept {
+                        return (address < other.address ? true : (address == other.address ? (tag < other.tag) : false));
+                    }
+
+                    // print
+                    // /** @brief print*/
+                    // template<class CharT, class Traits>
+                    // friend std::basic_ostream<CharT, Traits>& operator << (std::basic_ostream<CharT, Traits>& os, const extended_domain_id_type& dom_id) {
+                    //     os << "{id=" << dom_id.id << ", tag=" << dom_id.tag << ", rank=" << dom_id.mpi_rank << "}";
+                    //     return os;
+                    // }
+                    /** @brief print*/
+                    template<class CharT, class Traits>
+                    friend std::basic_ostream<CharT, Traits>& operator << (std::basic_ostream<CharT, Traits>& os, const extended_domain_id_type& dom_id) {
+                        os << "{tag=" << dom_id.tag << ", rank=" << dom_id.mpi_rank << "}";
+                        return os;
+                    }
+
+                };
+
+                // halo map type
+                using map_type = std::map<extended_domain_id_type, index_container_type>;
+
+                // static member functions
+
+                /** @brief compute number of elements in an object of type index_container_type */
+                static int num_elements(const index_container_type& c) noexcept {
+                    std::size_t s{0};
+                    for (const auto& is : c) s += is.size();
+                    return s;
+                }
+
+            private:
+
+                // members
+                communicator_type m_comm;
+                iteration_space_pair m_domain;
+                extended_domain_id_type m_id;
+                map_type m_send_map;
+                map_type m_recv_map;
+                pattern_container_type* m_container;
+
+            public:
+
+                // ctors
+                pattern(communicator_type& comm, const iteration_space_pair& domain, const extended_domain_id_type& id) :
+                    m_comm(comm),
+                    m_domain(domain),
+                    m_id(id) {}
```

Comment:
Where are the other members set? Expecially `m_container`, should it be set to nullptr?

Reply:
Yep, right. Done.